### ISSUE TITLE
Add UTF-8 encoding to Response header

### ISSUE THREAD

**Issue by matthieu-labas** [NanoHttpd/nanohttpd#261]
In `NanoHTTPD.newFixedLengthResponse()`, the `String` bytes are retrieved using `UTF-8` charset, but there is no explicit `UTF-8` charset specified in the response header, leading to incorrect display in browsers (which use the default encoding).

Adding `<meta http-equiv='Content-Type' content='text/html;charset=UTF-8'>` in the `<head>` element of my string corrects the problem (or `<meta charset="UTF-8">` for HTML5), but it seems to me that it should be handled in the response directly, as the encoding is performed there.

**Comment by LordFokas** [comment]
There's a problem here with the header. Unfortunately HTTP wasn't as thoroughly thought out as it should, so not everything is as easy as it should be. We have no right to mess with the served content, so our only option is to change the header. It actually makes sense that after we call `serve()` we set the charset to UTF-8. The problem is that header is 2 in one. Mime Type and Charset go packed in a single header, so it's difficult to change it because we might corrupt whatever the user has already put in there.

I suggest we check if the header exists. If not, we use "text/plain; charset=UTF-8".
If it exists we check if it starts with "text/" **and** it doesn't contain a ";" followed by "charset=". If the check fails, we let it be, else we append "; charset=UTF-8".

This is a nice chance to allow the programmer to set an Encoding in the Response object, defaulting to UTF8. Then the correct encoding is set on the header and used to extract the bytes from the String.

I'd like to hear from @ritchieGitHub or @elonen first but I might just do it and send a PR once I get home if I'm not terribly busy. **:)**

**Comment by matthieu-labas** [comment]
boolean charsetPresent(IHTTPSession session) {
		for (String content : session.getParms().get("Content-Type").split(";"))
			for (String param : content.split("="))
				if ("charset".equalsIgnoreCase(param))
					return true;
		return false;
	}

If it's messier than that (after adding the `null` checks), I'm not sure it's worth the trouble... As you said, it's basically HTTP's fault ;-)

**Comment by LordFokas** [comment]
I think the cleanest solution would be to create a ContentType class that takes a type string and an Encoding. Defaults would be, say, "application/octet-stream" & "UTF-8". After serve is called the correct header is compiled and overwritten. The right encoding would also be used to extract byte[] from strings. If the mime type doesn't start with "text/" everything mime type-related is ignored.

**Comment by ritchieGitHub** [comment]
I think @LordFokas  is right that there should be a class for that, there is already a method that will get the encoding from the content-type, but it is only used in the multi part case. I changed the code a little, now it should be more consistent. But still we should make this ContentType class.

**Comment by ritchieGitHub** [comment]
Did my fix help?

**Comment by ritchieGitHub** [comment]
closed, because no reaction

**Comment by matthieu-labas** [comment]
Sorry I was not able to test it until today (to be honnest, it was on my todo list that I forgot about).

I'm not sure I understand how to use the charset feature you added. Here is a small example of the problem I face, based on the version pulled today:

```
@Override public Response serve(IHTTPSession session) {
    return newFixedLengthResponse(Response.Status.OK, MIME_HTML, "<html><head><title>Testé ça</title></head><body>Testé ça</body></html>");
}
```
But I get `TestÃ© Ã§a` as an output, instead of `Testé ça`.

Unless I add `<meta http-equiv='Content-Type' content='text/html;charset=UTF-8'>`, I get the correct output.

What did your fix do in that regard? I probably missed something obvious.

Sorry about the delay, thank you for waking me up! :)

**Comment by ritchieGitHub** [comment]
@LordFokas you where right there was no way around a separate class for the content type, The code I wrote is not jet very efficient but it now checks if the passed text can be send using the specified encoding and if not it try's to use UTF-8. Including a unit test.

I did not cover the case of a user error if the encoding is in the mimeType and it can not encode the String properly. We could throw an exception...
